---
title: |
    | Low level Processing and Visualization
    | of Genome Sequencing Data
subtitle: "GSND 5340Q, BMDA"
author: | 
  | W. Evan Johnson, Ph.D.
  | Professor, Division of Infectious Disease
  | Director, Center for Data Science
  | Rutgers University -- New Jersey Medical School
date: "`r Sys.Date()`"
header-includes:
   - \usepackage{amsmath}
output: 
  beamer_presentation:
    theme: "CambridgeUS"
editor_options: 
  chunk_output_type: console
tables: true
---

# Motivating Example: X-linked Disease

## Rare X-linked Disease
![](figs/qc/proband.png)

## Rare X-linked Disease
![](figs/qc/family.png)


## Rare X-linked Disease
![](figs/qc/pedigree.png)


## Exon Capture Sequencing
![](figs/qc/hiseq_2000.jpg){width=40%}
![](figs/qc/illumina4.jpg){width=50%}

## Exome Capture
\center
![](figs/qc/exome1.png){height=90%}

## Exome Capture
\center
![](figs/qc/exome2.png){height=90%}

## Rare X-linked Disease
\center
![](figs/qc/seqPedigree.png){height=90%}

## Rare X-linked Disease
\center
![](figs/qc/snpfilter.pdf){height=80%}

## Rare X-linked Disease
![](figs/qc/cytoband.png)

## Rare X-linked Disease
![](figs/qc/ogdens_pub.png)


## N-terminal acetyltransferase (NAA10)
N-terminal acetyltransferase (NAT):

- Common modification (~80-90% of human proteins)
- Depletion from cancer cells linked to cell cycle arrest and apoptosis (Starheim, _BMC Proc_ 2009)
-	NAT genes directly implicated as cause of genetic disease
- Mutation demonstrated a significantly impaired biochemical activity _in vitro_
-	NAA10 lethal if knocked out of Drosophila

# Generating Sequencing Data

## Next-Generation Sequencing
- Expensive to purchase (hundreds of thousands $USD)
- Expensive to operate (e.g. reagents, flow cells)
- You can sequence your genome at 30X depth for <$1000 USD.

![](figs/qc/seq_machines.png)

## Next-Generation Sequencing
![](figs/qc/seq_process.png)


## Illumina Sequencing

- Most common sequencing technology today
- Sequences any DNA
- Sequencing by synthesis method
- Sequences (reads) are short (<300bp)
- 2 gigabases - 6 terabases per run
- Hours to days to complete one run

For more information, you can watch the following: 
[https://www.youtube.com/watch?v=fCd6B5HRaZ8](https://www.youtube.com/watch?v=fCd6B5HRaZ8)


## Illumina Sequencing
![](figs/qc/illumina_process.png)

## Sequencing Library Generation Workflow

Sequencing Library: DNA prepared for sequencing

Workflow:

- Extract RNA/DNA from sample
  - If RNA, reverse transcribe to cDNA
- Size select using gel cut or random shearing
- PCR amplify DNA if concentration is low
- Add sequencing adapters
  - If multiplexing, use barcoded adapters
- Pool samples, load across flow lanes for sequencing
- Typically only perform 1, sequencing cores 

## Design Choice: Single End vs Paired End
![](figs/qc/seq_design.png)

## Design Choice: Number of Reads
![](figs/qc/seq_design1.png)
[https://genohub.com/recommended-sequencing-coverage-by-application/](https://genohub.com/recommended-sequencing-coverage-by-application/)


## Design Choice: Sequencing Depth
Whole Exome

- Less expensive
- Nearly complete ascertainment of variation in the coding ~1\% of the genome (i.e. exons)
- Will miss functional variants outside of the coding region
	
Low Coverage Whole Genome 

- Less expensive
- Reasonably good ascertainment of shared variation, but not unique variation

Deep Whole Genome 

- More expensive
- Capture most of the genetic information
- Sequence the entire genome of each subject

## 3rd Gen Sequencing: Oxford Nanopore
![](figs/qc/nanopore.png)
[https://www.youtube.com/watch?v=GUb1TZvMWsw](https://www.youtube.com/watch?v=GUb1TZvMWsw)

<!--
## Analysis Pipelines (Broad Institute)
Queen/dictator of all analyses workflows: Genome Analysis Toolkit (GATK)
![](figs/qc/pipeline_overview-1.png)

[Link to GATK at the Broad](https://software.broadinstitute.org/gatk/)
-->

# Sequencing Data Formats

## Biological Data Formats are standardized

- Many different types of data
- Standard formats exist for many of them
- Always use/extend standard formats
- Donâ€™t save data in non standard formats when standards are available - No .xslx!
- For complete list of formats see: [https://genome.ucsc.edu/FAQ/FAQformat.html](https://genome.ucsc.edu/FAQ/FAQformat.html) 


## Sequencing Data Formats
![](figs/qc/seq_datatypes1.png)

## Sequencing Data Formats
![](figs/qc/seq_datatypes2.png)

## Sequencing Data Formats
![](figs/qc/seq_datatypes3.png)

## Data Formats Example
![](figs/qc/seq_example1.png)

## Data Formats Example
![](figs/qc/seq_example2.png)

## Data Formats Example
![](figs/qc/seq_example3.png)

## Data Formats Example
![](figs/qc/seq_example4.png)

## Data Formats Example
![](figs/qc/seq_example5.png)

## Data Formats Example
![](figs/qc/seq_example6.png)

## Data Formats Example
![](figs/qc/seq_example7.png)

## Data Formats Example
![](figs/qc/seq_example8.png)

## Data Formats Example
![](figs/qc/seq_example9.png)

## Data Formats Example
![](figs/qc/seq_example10.png)

## Data Formats Example
![](figs/qc/seq_example11.png)


# Raw Sequencing Data and QC

## History and Evolution of Illumina Data Output
Illumina sequencers have given output in many different formats:

- Illumina .PRB and .INT files
  - Better access to raw data. 
  - Base calling algorithms (Bravo and Irizarry, _Biometrics_, 2010)
  - Mapping algorithms (GNUMAP, NOVO)
  - Confusing formats; Large data files 		
- Illumina .FASTQ files
- Sanger .FASTQ files	
	
## Illumina .INT and .PRB
![](figs/qc/int_file.png){width=150%}

\vskip .2in
![](figs/qc/prb_file.png){width=150%}

## .fasta or .fa
![](figs/qc/fasta_example.png)

## .fastq or .fq (or .fq.gz)
![](figs/qc/illum_fastq2.png){width=125%}
\vskip .1in
![](figs/qc/illum_fastq1.png){width=125%}

## .FASTQ Paired End
![](figs/qc/fastq_paired.png)

## .FASTQ Comparison
![](figs/qc/fastq_comparison.png){width=100%, height=150%}
[https://en.wikipedia.org/wiki/Phred_quality_score](https://en.wikipedia.org/wiki/Phred_quality_score)

## Sequence Quality Score (PHRED)
![](figs/qc/phred.png)

[https://en.wikipedia.org/wiki/Phred_quality_score](https://en.wikipedia.org/wiki/Phred_quality_score)


## Sequence Quality Score (PHRED)
Quality scores (Phred)

- Sanger Phred: Range=(0,40), $P = 1-10^{-(ASCII-33)/10}$
- Solexa: Range= (-5,40),  $P = \frac{10^{(ASCII-64)/10)}}{ 1+10^{(ASCII-64)/10}}$
- Illumina 1.3: (0,40),  $P = \frac{10^{(ASCII-64)/10)}}{ 1+10^{(ASCII-64)/10}}$
- Illumina 1.5: Range=(2,40), $P = 1-10^{-(ASCII-64)/10}$
- Illumina 1.8: Same as Sanger except Range=(0,41)


## .FASTQ Comparison
![](figs/qc/fastq_comp.png)


<!--
## Preprocessing (Broad Institute)
\center
![](figs/qc/dznasg7toiq1.png)
}
-->

## Quality Control
Need to preprocess the reads:

- Check for quality (FASTQC)
-	Trim adapter and (Cutadapt, others)
- Remove duplicate reads, trim  low complexity/quality bases/reads (Prinseq)
- Complete pipelines: NCBI Toolkit, QC-Chain, PathoQC (pathoscope.sourceforge.net), others

\vskip .2in
**Note:** Not comprehensive or updated! 

## Read Quality Checks
\center
![](figs/qc/flowchart.png){height=90%}

## Read Quality Checks (Outdated!)
\center
![](figs/qc/methods.png)

## FASTQC
\center
![](figs/qc/fastqc.png){height=90%}






## FastQC Example
Running FastQC from command-line: 

```{bash, eval=F}	
fastqc myfastqfile.fq --outdir=output/
```

Interactive GUI:
```{bash, eval=F}	
fastqc
```


# Mapping Reads

## Sequencing data alignment
Next-Generation Sequencing (NGS) Data present new challenges:

- Map `reads' to genome
- Call SNPs and variants from the reads
- Other Applications: RNA-seq, miRNAs, alternative splicing, ChIP-seq, BS-seq, RNA editing

Obstacles:

- Massive data size
- Repeat regions, rare variants
	
\vskip .2in
**Goal:** Develop an approach to put the puzzle together! 

<!--
%\frame{
%\frametitle{Mapping Algorithms}
%	\begin{itemize}
%		\item BLAST
%		\item BLAT
%		\item SOAP2
%		\item Bowtie2
%		\item BWA
%		\item STAR
%		\item Subread (Rsubread)
%		\item GNUMAP
%		\item NovoAlign
%		\item Many others!!
%	\end{itemize}
%}
-->

## Basic Local Alignment Search Tool (BLAST)
BLAST seed extension:
![](figs/qc/Neighbor_HSP.jpg)

## Creating a Hash Index Table
![](figs/qc/hash1.png)

## Creating a Hash Index Table
![](figs/qc/hash2.png)

## Creating a Hash Index Table
![](figs/qc/hash3.png)

## Creating a Hash Index Table
![](figs/qc/hash4.png)

## Creating a Hash Index Table
![](figs/qc/hash5.png)

## Read Lookup
![](figs/qc/read1.png)

## Read Lookup
![](figs/qc/read2.png)

## Read Lookup
![](figs/qc/read3.png)

## Read Lookup
![](figs/qc/read4.png)

## Read Lookup
![](figs/qc/read5.png)

## Needleman-Wunsch
![](figs/qc/needleman_wunsch1.png){width=85%}

## Needleman-Wunsch
![](figs/qc/needleman_wunsch2.png){width=85%}

## Needleman-Wunsch
![](figs/qc/needleman_wunsch3.png){width=85%}

## Needleman-Wunsch
![](figs/qc/needleman_wunsch4.png){width=85%}

## Prefix Trees
\center
![](figs/qc/trie.png)


## Prefix Trees
\center
![](figs/qc/bwatrie.png){height=90%}


## Burrows-Wheeler Transformation (BWA, Bowtie)
\center
![](figs/qc/bwa.jpg){height=85%}

## Burrows-Wheeler Transformation (BWA, Bowtie)
\center
![](figs/qc/bowtie.pdf)

## Hybrid Methods (BWA-SW and MEM, Bowtie2)
\center
![](figs/qc/bowtie2flow.jpg)

#Hybrid Methods (BWA-SW, Bowtie2)
\center
![](figs/qc/bowtie2flow.jpg){height=80%}


<!--
\begin{frame}[fragile]
	\frametitle{BWA Example}
	\begin{enumerate}
	\item Index the genome:
	\begin{verbatim}
	bwa index genomefile.fa 
	\end{verbatim}
	\item Align the reads$^*$:
	\begin{verbatim}
	bwa mem genomefile.fa myfastqfile.fq > bwaalign.sam
	\end{verbatim}
	\end{enumerate}
	
	\vskip .2in
	$^*$Note: the bwa mem should include a `read group' header as well, e.g.: 
	\small
	\begin{verbatim}
	-r @RG\tID:group1\tSM:sample1\tPL:illumina\tLB:lib1\tPU:unit1 
	\end{verbatim}

\end{frame}
	
	
	\begin{frame}[fragile]
	\frametitle{Bowtie2 Example}
	\begin{enumerate}
	\item Index the genome:
	\begin{verbatim}
	bowtie2-build genomefile.fa output/genomefile	
	\end{verbatim}
	\item Align the reads:
	\begin{verbatim}
	bowtie2 -x output/genomefile -U myfastqfile.fq \
	-S bowtiealign.sam
	\end{verbatim}
	\end{enumerate}
\end{frame}
-->	



# Visualization of Aligned Sequencing Data

## Options for data visualization

- UCSC Genome Browser
- IGB (http://bioviz.org/igb/)
- IGV (http://www.broadinstitute.org/igv/)
- SAMtools


## IGB Example
\center
![](figs/qc/beforeExtContig.png)

## IGV Example
\center
![](figs/qc/igv_example.png){height=85%}

## SAM Format
\center
![](figs/qc/samformat.png)

## SAMtools
\center
![](figs/qc/samtools.png)
<!--
\begin{frame}[fragile]
	\frametitle{SAMtools Example}
	\begin{enumerate}
	\item Index the genome:
	\begin{verbatim}
	samtools faidx genomefile.fa
	\end{verbatim}
	\item Convert SAM file to BAM format:
	\begin{verbatim}
	samtools import genomefile.fa.fai myalignments.sam \
	myalignments.bam 
	\end{verbatim}
	\item Sort the BAM file:
	\begin{verbatim}
	samtools sort myalignments.bam -o myalignments.sorted.bam
	\end{verbatim}
	%samtools merge -f merged.bam *sorted.bam # Merge BAM files--note not necessary here
	\item Index the BAM file:
	\begin{verbatim}
	samtools index myalignments.sorted.bam 
	\end{verbatim}
	\item View BAM file in SAMtools:
	\begin{verbatim}
	samtools tview myalignments.sorted.bam genomefile.fa
	\end{verbatim}
	\end{enumerate}
\end{frame}


\begin{frame}[fragile]	
	\frametitle{SAMtools Example}
	My fast and simple script for SAMtools preparation:
	\begin{verbatim}
	python prepsam.py genomefile.fa myalignments.sam
	\end{verbatim}
	
	\vskip .1in
	The script can also prepare and merge multiple .sam files (e.g.):
	\begin{verbatim}
	python prepsam.py genomefile.fa sam1.sam sam2.sam
	python prepsam.py genomefile.fa *.sam
	\end{verbatim}


\end{frame}

\frame
{
	\frametitle{SAMtools Example}
	SAMtools commands:
	\begin{itemize}
		\item The one command to remember: `?'
		\item g: go to a specific location (i.e. chrX:152852988 or chrX\_149913753:2939235)
		\item m,b: mapping quality, base quality
		\item n: color by nucleotide
		\item `.': dot/base view
		\item r: read name
		\item q: quit SAMtools
	\end{itemize}

	% python prepsam.py genome/chrX_5MB.fa map_gnu/proband_short.sam
}
-->


# Other Data Formats

## Standard format for keeping tables
![](figs/qc/tables.png)

## Genomic regions

-A region is defined by three required fields:
  - sequence name (e.g. chromosome)
  - start coordinate
  - end coordinate
- Define regions of interest: introns, exons, genes, etc.
- Additional information saved as fields after the first three.
- Three standard tab-separated formats: BED, GFF, GTF
No headers

## BED Format
Mandatory fields:

1. chrom - Name of the chromosome/scaffold/reference sequence
2. chromStart - 0-based starting position of the feature on chrom
3. chromEnd - Ending position of the feature in the chromosome or scaffold.

The chromEnd base is not included in the display of the feature.

For example, the first 100 bases of a chromosome are defined as:

- chromStart=0
- chromEnd=100
- span the bases numbered 0-99

## BED Format
![](figs/qc/bed_format1.png)

## BED Format
![](figs/qc/bed_format2.png)


## BED Format
Optional fields:

4. Name
5. Score
6. Strand
7. 7-12. Display options (thick starts and end, color, blocksâ€¦) to control the view on the genome browser

## bedtools
[http://bedtools.readthedocs.io/](http://bedtools.readthedocs.io/)

- sort (sort bed files)
- Intersect (get intersections of bed files)
- merge
- coverage
- overlap
- substract
- ...

## bedtools intersect
![](figs/qc/bedtools1.png)

## bedtools merge
![](figs/qc/bedtools2.png)

## GFF - General features

9 mandatory fields, tab separated:

1. seqname - The name of the sequence. Must be a chromosome or scaffold.
2. source - The program that generated this feature.
3. feature - The name of this type of feature (e.g. gene, exon, etc). 
4. start - The starting position of the feature in the sequence (1-based)
5. end - The ending position of the feature (inclusive).
6. score - A score between 0 and 1000. 
7. strand - Valid entries include "+", "-", or "." (for don't know/don't care).
8. frame - If the feature is a coding exon, frame should be a number between 0-2 that represents the reading frame of the first base. If the feature is not a coding exon, the value should be ".".
9. group - All lines with the same group are linked together into a single item

## Gene Information
GTF (Gene Transfer Format, GTF2.2)

- Extension to GFF2, backwards compatible
- First eight GTF fields are the same as GFF
- _feature field_ is the same as GFF, has controlled vocabulary:
  - _gene, transcript, exon, CDS, 5UTR, 3UTR, inter, inter_CNS, and intron_CNS, etc_
- group field expanded into a list of attributes (i.e. key/value pairs)

The attribute list must begin with the one mandatory attribute:
  - gene_id value - A globally unique identifier for the genomic source of the sequence


## GTF format
![](figs/qc/gtf.png)


<!--

\section{SNP Calling}
\subsection{Methods for SNP Calling }
\frame
{
	\frametitle{Methods for SNP Calling }
	Methods for SNP calling:
	\begin{itemize}
		\item Mapper/callers: MAQ, SOAPsnp, GNUMAP, Crossbow (Bowtie)
		\item Callers: SAMtools (mpileup), GATK (HaplotypeCaller, Mutect2), FreeBayes, others
	\end{itemize}	
	}



\begin{frame}{Inconsistencies Among Aligners}
Low concordance of variant-calling pipelines (O'Rawe, {\it Genome Med}, 2013)
\begin{center}
	\includegraphics[height=2.8in]{figures/discordance}
\end{center}
\end{frame}



\begin{frame}[fragile]	
	\frametitle{SAMtools Example}
	\begin{itemize}
	\item Multiple sample SNP calling:
	\begin{verbatim}
	samtools mpileup -f genomefile.fa \ 
	myalignments.sorted.bam > myalignments.vcf
	\end{verbatim}
	\end{itemize}
\end{frame}


\begin{frame}[fragile]	
	\frametitle{GATK Example}
	
	\begin{verbatim}
	java -jar CreateSequenceDictionary.jar \
	R=chrX_5MB.fa O=chrX_5MB.dict
	\end{verbatim}
	
	\vskip .05in
	
	\begin{verbatim}
	gatk --java-options "-Xmx8G"  HaplotypeCaller \
	-R chrX_5MB.fa -I proband_short_bwa.sorted.bam \
	-O proband_short.vcf
	\end{verbatim}
	
	
\end{frame}


\frame
{
	\frametitle{.vcf files}
			\includegraphics[width=4.75in]{figures/vcf_format}
}



\frame
{
	\frametitle{Complete Variant Calling Pipeline  (Outdated!)}
	Our analysis pipeline consisted of the following:
	\begin{enumerate}
		\item Align the FASTQ files to genome
		\item Convert SAM file to BAM, add read group info
		\item Filter the reads based on quality (BAMTools)
		\item Samtools to sort and index, and use Picard to mark duplicates
		\item GATK calibration, realignment, variant calling (HaplotypeCaller, Mutect2)
		\item Filter the called variants (GATK filtersnps and filterindels).
		\item Annotation of SNPs (snpEff, condel)
		\item Filter by frequency (thousand genomes, TCGA, etc.)
		\item Downstream analysis (rare variants, pedigree, pathway level, etc)		
	\end{enumerate}
}






\subsection{Downstream Annotation and Analysis}
\frame
{
	\frametitle{Downstream Annotation and Analysis (Outdated!)}
	Downstream Annotation Tools (old list):
	\begin{itemize}
	 	\item snpEff (http://snpeff.sourceforge.net/)
		\item Condel (http://bg.upf.edu/condel/home)
			\begin{itemize}
			\item SIFT http://sift.jcvi.org/
			\item Polyphen 2 http://genetics.bwh.harvard.edu/pph2/
			\item http://mutationassessor.org/ 
			\end{itemize}
		\item Ensembl variant effect predictor (http://www.ensembl.org/info/docs/variation/vep/index.html)
		\item Thousand Genomes variant frequency (e.g. 1\% threshold) and Exome Sequence Project variant frequency (e.g. 1\%).
	\end{itemize}

	% python prepsam.py genome/chrX_5MB.fa map_gnu/proband_short.sam
}

-->

## Session info
\tiny
```{r session info}
sessionInfo()
```

